cmd/codegen/generator/go/templates/module_funcs.go

```diff
@@ -0,0 +1,337 @@
+				}
+				spec.parent = parent
+				specs = append(specs, spec)
+			}
+			return specs, nil
+		}
+	}
+
+	// if other parameter passing schemes fail, just treat each remaining arg
+	// as a top-level param
+	paramFields := unpackASTFields(fnDecl.Type.Params)
+	for ; i < params.Len(); i++ {
+		docComment, lineComment := ps.commentForFuncField(fnDecl, paramFields, i)
+		spec, err := ps.parseParamSpecVar(params.At(i), docComment.Text(), lineComment.Text())
+		if err != nil {
+			return nil, err
+		}
+		if sig.Variadic() && i == params.Len()-1 {
+			spec.variadic = true
+		}
+		specs = append(specs, spec)
+	}
+	return specs, nil
+}
+
+func (ps *parseState) parseParamSpecVar(field *types.Var, docComment string, lineComment string) (paramSpec, error) {
+	if _, ok := field.Type().(*types.Struct); ok {
+		return paramSpec{}, fmt.Errorf("nested structs are not supported")
+	}
+
+	paramType := field.Type()
+	baseType := paramType
+	for {
+		ptr, ok := baseType.(*types.Pointer)
+		if !ok {
+			break
+		}
+		baseType = ptr.Elem()
+	}
+
+	optional := false
+	defaultValue := ""
+
+	if named, ok := ps.isOptionalWrapper(baseType); ok {
+		typeArgs := named.TypeArgs()
+		if typeArgs.Len() != 1 {
+			return paramSpec{}, fmt.Errorf("optional type must have exactly one type argument")
+		}
+		optional = true
+
+		baseType = typeArgs.At(0)
+		for {
+			ptr, ok := baseType.(*types.Pointer)
+			if !ok {
+				break
+			}
+			baseType = ptr.Elem()
+		}
+	}
+
+	docPragmas, docComment := parsePragmaComment(docComment)
+	linePragmas, lineComment := parsePragmaComment(lineComment)
+	comment := strings.TrimSpace(docComment)
+	if comment == "" {
+		comment = strings.TrimSpace(lineComment)
+	}
+
+	pragmas := make(map[string]string)
+	maps.Copy(pragmas, docPragmas)
+	maps.Copy(pragmas, linePragmas)
+	if v, ok := pragmas["default"]; ok {
+		defaultValue = v
+	}
+	if v, ok := pragmas["optional"]; ok {
+		if v == "" {
+			optional = true
+		} else {
+			optional, _ = strconv.ParseBool(v)
+		}
+	}
+
+	// ignore ctx arg for parsing type reference
+	var typeSpec ParsedType
+	if paramType.String() != contextTypename {
+		var err error
+		typeSpec, err = ps.parseGoTypeReference(baseType, nil)
+		if err != nil {
+			return paramSpec{}, fmt.Errorf("failed to parse type reference: %w", err)
+		}
+	}
+
+	return paramSpec{
+		name:         field.Name(),
+		paramType:    paramType,
+		baseType:     baseType,
+		typeSpec:     typeSpec,
+		optional:     optional,
+		defaultValue: defaultValue,
+		description:  comment,
+	}, nil
+}
+
+type paramSpec struct {
+	name        string
+	description string
+
+	optional bool
+	variadic bool
+
+	defaultValue string
+
+	// paramType is the full type declared in the function signature, which may
+	// include pointer types, Optional, etc
+	paramType types.Type
+	// baseType is the simplified base type derived from the function signature
+	baseType types.Type
+	// typeSpec is the parsed TypeSpec of the argument's baseType
+	typeSpec ParsedType
```

Comment:
With this, we might be able to remove `baseType` from here completely which would be nice.

All we'd need to do would be to add a new method to `ParseType` that returns the `type.Type` of that type (currently it's only possible to get the children).